client/asset/dcr/dcr.go

```diff
@@ -4521,3 +4540,190 @@ func float64PtrStr(v *float64) string {
 	}
 	return strconv.FormatFloat(*v, 'f', 8, 64)
 }
+
+func (dcr *ExchangeWallet) ConfirmRedemption(coinID dex.Bytes, redemption *asset.Redemption, feeSuggestion uint64) (*asset.ConfirmRedemptionStatus, error) {
+	txHash, _, err := decodeCoinID(coinID)
+	if err != nil {
+		return nil, err
+	}
+
+	var secretHash [32]byte
+	copy(secretHash[:], redemption.Spends.SecretHash)
+	dcr.mempoolRedeemsMtx.RLock()
+	mRedeem, have := dcr.mempoolRedeems[secretHash]
+	dcr.mempoolRedeemsMtx.RUnlock()
+
+	var deleteMempoolRedeem bool
+	defer func() {
+		if deleteMempoolRedeem {
+			dcr.mempoolRedeemsMtx.Lock()
+			delete(dcr.mempoolRedeems, secretHash)
+			dcr.mempoolRedeemsMtx.Unlock()
+		}
+	}()
+
+	var abandonTx bool
+	tx, err := dcr.wallet.GetTransaction(dcr.ctx, txHash)
+	if err == nil {
+		if have && mRedeem.txHash == *txHash {
+			if tx.Confirmations == 0 && time.Now().After(mRedeem.firstSeen.Add(maxRedeemMempoolAge)) {
+				// Transaction has been sitting in the mempool
+				// for a long time now. Attempt to abandon.
+				abandonTx = true
+				if err := dcr.wallet.AbandonTransaction(dcr.ctx, txHash); err != nil {
+					dcr.log.Warnf("Unable to abandon redemption mempool transaction %v: %v.", txHash, err)
+				} else {
+					dcr.log.Infof("Abandoned redemption mempool transaction: %v.", txHash)
+					deleteMempoolRedeem = true
+				}
+			}
+		} else {
+			// Will hit this if node was restarted with an actively
+			// redeeming swap.
+			dcr.mempoolRedeemsMtx.Lock()
+			dcr.mempoolRedeems[secretHash] = &mempoolRedeem{txHash: *txHash, firstSeen: time.Now()}
+			dcr.mempoolRedeemsMtx.Unlock()
+		}
+		if !abandonTx {
+			if tx.Confirmations >= requiredRedeemConfirms {
+				deleteMempoolRedeem = true
+			}
+			return &asset.ConfirmRedemptionStatus{
+				Confs:  uint64(tx.Confirmations),
+				Req:    requiredRedeemConfirms,
+				CoinID: coinID,
+			}, nil
+		}
+	}
+
+	if !abandonTx && !errors.Is(err, asset.CoinNotFoundError) {
```

Comment:
It's a bit challenging to reason out if it's possible to have a `nil` `err` and `abantonTx=false`.  Can that happen?
I keep thinking that before this it should do:

```go
if err == nil {
    if have && ...
    ...
} else if !errors.Is(err, asset.CoinNotFoundError) {
    return ...
}

// We are here if the tx not found, or found but abandoning...
````
?

Reply:
I'm not sure what I was thinking here. I moved `!errors.Is(err, asset.CoinNotFoundError)` up because if it is not that error we should fail anyway.